gstreamer/src/query.rs

```diff
@@ -208,84 +208,273 @@ impl fmt::Debug for QueryRef {
+// `Query` can also be accessed as `&mut` thanks to the controls
+// enforced by `QueryWrapper`
+impl DerefMut for QueryView {
+    fn deref_mut(&mut self) -> &mut Query {
+        match *self {
+            QueryView::Position(ref mut position) => position,
+            QueryView::Duration(ref mut duration) => duration,
+            QueryView::Latency(ref mut latency) => latency,
+            QueryView::Jitter(ref mut jitter) => jitter,
+            QueryView::Rate(ref mut rate) => rate,
+            QueryView::Seeking(ref mut seeking) => seeking,
+            QueryView::Segment(ref mut segment) => segment,
+            QueryView::Convert(ref mut convert) => convert,
+            QueryView::Formats(ref mut formats) => formats,
+            QueryView::Buffering(ref mut buffering) => buffering,
+            QueryView::Custom(ref mut custom) => custom,
+            QueryView::Uri(ref mut uri) => uri,
+            QueryView::Allocation(ref mut allocation) => allocation,
+            QueryView::Scheduling(ref mut scheduling) => scheduling,
+            QueryView::AcceptCaps(ref mut accept_caps) => accept_caps,
+            QueryView::Caps(ref mut caps) => caps,
+            QueryView::Drain(ref mut drain) => drain,
+            QueryView::Context(ref mut context) => context,
+            QueryView::Other(ref mut other) => other,
+            _ => unimplemented!("Converting QueryView into Query for unknown query type"),
+        }
+    }
+}
+
+impl<'a> From<&'a QueryView> for &'a QueryRc {
+    fn from(view: &'a QueryView) -> Self {
+        match *view {
+            QueryView::Position(ref position) => &position.0,
+            QueryView::Duration(ref duration) => &duration.0,
+            QueryView::Latency(ref latency) => &latency.0,
+            QueryView::Jitter(ref jitter) => &jitter.0,
+            QueryView::Rate(ref rate) => &rate.0,
+            QueryView::Seeking(ref seeking) => &seeking.0,
+            QueryView::Segment(ref segment) => &segment.0,
+            QueryView::Convert(ref convert) => &convert.0,
+            QueryView::Formats(ref formats) => &formats.0,
+            QueryView::Buffering(ref buffering) => &buffering.0,
+            QueryView::Custom(ref custom) => &custom.0,
+            QueryView::Uri(ref uri) => &uri.0,
+            QueryView::Allocation(ref allocation) => &allocation.0,
+            QueryView::Scheduling(ref scheduling) => &scheduling.0,
+            QueryView::AcceptCaps(ref accept_caps) => &accept_caps.0,
+            QueryView::Caps(ref caps) => &caps.0,
+            QueryView::Drain(ref drain) => &drain.0,
+            QueryView::Context(ref context) => &context.0,
+            QueryView::Other(ref other) => &other.0,
+            _ => unimplemented!("Converting QueryView into Query for unknown query type"),
+        }
+    }
+}
+
+impl From<QueryView> for QueryWrapper {
+    fn from(query_view: QueryView) -> Self {
+        QueryWrapper(query_view)
+    }
+}
+
+pub struct QueryWrapper(QueryView);
```

Comment:
What's the actual difference between a `QueryWrapper` and a `GstRc<QueryRef>`? The below is basically the `Arc` API. And from it you can get the view and the `&QueryRef`.

I don't understand why those two types can't be the same, I'm probably missing something

Reply:
In my perspective, `QueryWrapper` is acting as `GstRc<T>`. The differences are:

- The position in the architecture: `QueryWrapper` is at the top level of the ownership chain, when `GstRc<T>` can be owned by a concrete type which can be owned by a `QueryView`.
- `GstRc<T>`'s parameter `T` is expected to be tied to a `ffi::type` which prevented me from using it to host a `QueryView`.

I kept `GstRc<T>` in the composition chain as a convenience for reusing existing functions. I should clear that by implementing them on the `QueryWrapper` and get it out of the way. Clearly, they provide the same functionalities. If we adopted this approach, we could use `GstRc<T>` in place of the `QueryWrapper`. I just didn't want to change all the code of `gstreamer-rs` for this architecture preview :)